### PR TITLE
Use locateFile in the dynamic module loader

### LINKED ISSUE #14502: locateFile not used when loading SIDE_MODULES
Experimenting with [locateFile](https://emscripten.org/docs/api_reference/module.html#Module.locateFile) for browser cache invalidation and noticed that SIDE_MODULES specified as `Module['dynamicLibraries'] = ['a.wasm', 'b.wasm']` do not pass through `locateFile` . The HTTP requests are missing `?v=revision`. HTTP requests for `app.wasm` and `app.data` have the revision string appended as expected. 

```JavaScript
Module['locateFile'] = function (path, prefix) {
  if (path.endsWith('.wasm') || path.endsWith('.data')) {
    return prefix + path + '?v=' + window.REVISION;
  }
  return prefix + path;
}
```
Noticed in 2.0.16 and confirmed that it also occurs in 2.0.20. I attempted to test later versions but some of the recent compiler/linker flag changes are preventing the project from compiling.

Is `locateFile` an appropriate tool? Is this a bug in the loading of SIDE_MODULES?

### PR DESCRIPTION
**Pull request by georgestagg**
With this change, in `loadLibData` the path to `libFile` is passed through `locateFile` before attempting to download the data from an external source. Fixes #14502.

A test for dynamic linking with `locateFile` is added that creates the `liblib.so` side module in a temporary directory rather than in the current directory, as default. The side module is linked by passing the full file path to `liblib.so`, and `locateFile` is then used to tell Emscripten at runtime where to find the file. The test only passes when the dynamic module loader is indeed able to use `locateFile` to find the side module for loading.

### CODE REVIEW

File: test/test_core.py [hunk 1]
```diff
@@ -4177,6 +4179,21 @@ def test_dylink_safe_heap(self):
     self.set_setting('SAFE_HEAP')
     self.do_basic_dylink_test()
 
+  @needs_dylink
+  def test_dylink_locate_file(self):
+    name = 'liblib.so'
+    with tempfile.TemporaryDirectory() as tmpdir:
```
Comment (sbc100) [inline test/test_core.py:4185]: I don't think there is any need to use a random tmpdir here.

How about just doing:`

```
so_dir = 'so_dir'
os.makedir(so_dir)
```

Each test already runs inside its own temporary sandbox directory anyway

File: test/test_core.py [hunk 2]
```diff
@@ -4177,6 +4179,21 @@ def test_dylink_safe_heap(self):
     self.set_setting('SAFE_HEAP')
     self.do_basic_dylink_test()
 
+  @needs_dylink
+  def test_dylink_locate_file(self):
+    name = 'liblib.so'
```
Comment (sbc100) [inline test/test_core.py:4184]: Maybe call this `so_name` for consistency?

**Review by sbc100 — COMMENTED** [review]
Great!  Thanks for PR.  Just a couple of nits in the test code.

File: test/test_core.py (hunk 1 above)
Comment (georgestagg) [reply test/test_core.py]: Sounds good to me. I was trying to clean up after myself but I can see now that tests run in a tmpdir anyway.

File: test/test_core.py (hunk 2 above)
Comment (georgestagg) [reply test/test_core.py]: Done. I've rebased with the changes.

File: test/test_core.py [hunk 3]
```diff
@@ -3102,7 +3102,7 @@ def test_dlfcn_missing(self):
     if self.js_engines == [config.V8_ENGINE]:
       expected = "error: Could not load dynamic lib: libfoo.so\nError: Error reading file"
     else:
-      expected = "error: Could not load dynamic lib: libfoo.so\nError: ENOENT: no such file or directory, open 'libfoo.so'"
+      expected = "error: Could not load dynamic lib: libfoo.so\nError: ENOENT: no such file or directory"
```
Comment (sbc100) [inline test/test_core.py:3105]: Is this change needed?  Seems like it would be good to check the filename appears in the error message.

**Review by sbc100 — APPROVED** [review]

File: test/test_core.py (hunk 3 above)
Comment (georgestagg) [reply test/test_core.py:3105]: Without the change the test fails because the full path including the tmp dir appears in the error message, e.g.
```
 error: Could not load dynamic lib: libfoo.so
-Error: ENOENT: no such file or directory, open 'libfoo.so'
+Error: ENOENT: no such file or directory, open '/private/var/folders/wx/qv_075m16yx5lqwvjfwswfqm0000gn/T/emscripten_test_core0_jwfom28o/libfoo.so'
```

I considered using something like,
```
expected = "error: Could not load dynamic lib: libfoo.so\nError: ENOENT: no such file or directory, open '.*libfoo.so'"
```
with
```
self.do_run(src, expected, regex=True)
```
but decided against it. Would you prefer that?

Comment (sbc100) [reply test/test_core.py:3105]: Perhaps we should use the original name in the error message then?

Comment (georgestagg) [reply test/test_core.py:3105]: On a second look, it seems the first line of the error already does include the original filename. Since the name also appears in the first part of the error message, perhaps that is enough?

I think the problem is that the exception `e` is directly added as the next line, and I'm not sure how much control there is over how that is printed.

https://github.com/emscripten-core/emscripten/blob/aab32efdcc9d1157fe925678ae5806a8f371d4ec/src/library_dylink.js#L997

Comment (sbc100) [reply test/test_core.py:3105]: OK fair enough.
